README
ICMP host name utilities

Since the Linux kernel does not support the ICMP host name facilities as
described in RFC 1788, I wrote a simple implementation in user-space.
It comprises three parts:

    * icmpdnd - The ICMP Domain Name daemon, which listens to ICMP
      domain name requests and replies to them.

    * idnlookup - A user utility to query a host for its host name over
      ICMP. Like ping and all other utilities that use ICMP, it needs to
      be installed SUID root.

    * libnss_icmp.so.2 - A glibc NSS module, which uses idnlookup to
      perform queries for the gethostbyaddr function.

General information

The NSS module caches all requests (including failures) and honors the
TTL value sent by the responding host. It uses a configuration file called
/etc/nss-icmp.conf, which controls different aspects of its operation:

    * timeout - Sets the timeout in milliseconds for ICMP host name
      queries, by passing the -t option to idnlookup.
    * nocache - Disables the cache.
    * ttlnotfound - Sets the TTL for not-found cache entries is
      seconds. The default is 5 minutes.

Both icmpdnd and idnlookup need to run as root, due to the fact that
they use ICMP. Thus, idnlookup needs be installed SUID root.



ChangeLog
2006-01-15  Fredrik Tolf  <fredrik@dolda2000.com>
	* configure.ac: Bump version to 0.4
	* admin/icmpdnd: Fixed chkconfig values to ensure that icmpdnd is
	started after ypbind.

2006-01-13  Fredrik Tolf  <fredrik@dolda2000.com>
	* src/*.[1-9]: Added man pages

2005-06-27  Fredrik Tolf  <fredrik@dolda2000.com>
	* src/icmpdnd.c: Made it possible to set the name explicitly.
	* configure.ac: Advance version to 0.3

2005-05-25  Fredrik Tolf  <fredrik@dolda2000.com>

	* admin/Makefile.am: Don't install the RedHat rc script on all systems.
	* configure.ac: Change version to 0.2


src/icmpdnd.c
/*
 *  icmpdnd - ICMP Domain Name responder daemon for Linux
 *  Copyright (C) 2005 Fredrik Tolf <fredrik@dolda2000.com>
 *  
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <syslog.h>
#include <time.h>
#include <sys/socket.h>
#include <sys/poll.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet/ip.h>
#include <sys/types.h>
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef MAXHNAME
#define MAXHNAME 1024
#endif

struct icmphdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
};

struct reqhdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
    u_int16_t id;
    u_int16_t seq;
};

struct rephdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
    u_int16_t id;
    u_int16_t seq;
    int32_t ttl;
};

#define ICMP_NAMEREQ 37
#define ICMP_NAMEREP 38

volatile int alive;
char myname[MAXHNAME] = "";

void setname(char *newname)
{
    int nl;
    
    if(newname == NULL) {
	if(gethostname(myname, sizeof(myname)) < 0) {
	    perror("gethostname");
	    exit(1);
	}
	nl = strlen(myname);
	myname[nl++] = '.';
	if(getdomainname(myname + nl, sizeof(myname) - nl) < 0) {
	    perror("getdomainname");
	    exit(1);
	}
	if(strlen(myname + nl) != 0) {
	    nl = strlen(myname);
	    myname[nl++] = '.';
	}
	myname[nl] = 0;
    } else {
	strcpy(myname, newname);
	nl = strlen(myname);
	if(myname[nl - 1] != '.') {
	    myname[nl] = '.';
	    myname[nl + 1] = 0;
	}
    }
}

size_t filldn(char *dst)
{
    char *p, *p2, *dp;
    char namebuf[MAXHNAME];
    
    strcpy(namebuf, myname);
    p = namebuf;
    dp = dst;
    while((p2 = strchr(p, '.')) != NULL) {
	*p2 = 0;
	*(dp++) = p2 - p;
	memcpy(dp, p, p2 - p);
	dp += p2 - p;
	p = p2 + 1;
    }
    *(dp++) = 0;
    
    return(dp - dst);
}

void cksum(void *hdr, size_t len)
{
    struct icmphdr *ih;
    u_int8_t *cb;
    int i;
    int b1, b2;
    
    ih = (struct icmphdr *)hdr;
    cb = (u_int8_t *)hdr;
    ih->checksum = 0;
    b1 = b2 = 0;
    for(i = 0; i < (len & ~1); i += 2) {
	b1 += cb[i];
	b2 += cb[i + 1];
    }
    if(i & 1)
	b1 += cb[len - 1];
    while(1) {
	if(b1 >= 256) {
	    b2 += b1 >> 8;
	    b1 &= 0xff;
	    continue;
	}
	if(b2 >= 256) {
	    b1 += b2 >> 8;
	    b2 &= 0xff;
	    continue;
	}
	break;
    }
    cb = (u_int8_t *)&ih->checksum;
    cb[0] = ~(u_int8_t)b1;
    cb[1] = ~(u_int8_t)b2;
}

int main(int argc, char **argv)
{
    int i, n, ret;
    int c, cs, s4, s6, datalen;
    int daemonize, ttl;
    unsigned char buf[65536];
    struct sockaddr_storage name;
    struct reqhdr req;
    struct rephdr rep;
    struct iphdr iphdr;
    struct msghdr mhdr;
    struct iovec iov;
    char cmsgbuf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
    size_t hdrlen;
    struct pollfd pfd[2];
    time_t curtime, lasterr;
    
    daemonize = 1;
    ttl = 3600;
    while((c = getopt(argc, argv, "nht:d:")) != -1) {
	switch(c) {
	case 't':
	    ttl = atoi(optarg);
	    break;
	case 'd':
	    setname(optarg);
	    break;
	case 'n':
	    daemonize = 0;
	    break;
	case 'h':
	case '?':
	case ':':
	default:
	    fprintf(stderr, "usage: icmpdnd [-nh] [-t ttl] [-d domainname]\n");
	    exit((c == 'h')?0:1);
	}
    }
    if(*myname == 0)
	setname(NULL);
    
    s4 = socket(PF_INET, SOCK_RAW, IPPROTO_ICMP);
    s6 = socket(PF_INET6, SOCK_RAW, IPPROTO_ICMP);
    if((s4 < 0) && (s6 < 0)) {
	perror("could not open raw socket");
	exit(1);
    }
    if(s6 >= 0) {
	i = 1;
	if(setsockopt(s6, IPPROTO_IPV6, IPV6_PKTINFO, &i, sizeof(i))) {
	    perror("could not set IPV6_PKTINFO sockopt");
	    exit(1);
	}
    }
    
    if(daemonize)
	daemon(0, 0);
    
    openlog("icmpdnd", LOG_PID | LOG_NDELAY, LOG_DAEMON);
    
    alive = 1;
    lasterr = 0;
    while(alive) {
	n = 0;
	if(s4 >= 0) {
	    pfd[n].fd = s4;
	    pfd[n].events = POLLIN;
	    n++;
	}
	if(s6 >= 0) {
	    pfd[n].fd = s6;
	    pfd[n].events = POLLIN;
	    n++;
	}
	ret = poll(pfd, n, -1);

	curtime = time(NULL);
	if(ret < 0) {
	    if(errno == EINTR)
		continue;
	    syslog(LOG_ERR, "error while polling sockets: %m");
	    if(lasterr == curtime) {
		syslog(LOG_CRIT, "exiting due to repeated errors");
		exit(1);
	    }
	    lasterr = curtime;
	}
	
	for(i = 0; i < n; i++) {
	    if((pfd[i].revents & POLLIN) == 0)
		continue;
	    cs = pfd[i].fd;
	    memset(&name, 0, sizeof(name));
	    
	    iov.iov_len = sizeof(buf);
	    iov.iov_base = buf;
	    mhdr.msg_name = &name;
	    mhdr.msg_namelen = sizeof(name);
	    mhdr.msg_iov = &iov;
	    mhdr.msg_iovlen = 1;
	    mhdr.msg_control = cmsgbuf;
	    mhdr.msg_controllen = sizeof(cmsgbuf);
	    
	    ret = recvmsg(cs, &mhdr, 0);
	    if(ret < 0) {
		syslog(LOG_WARNING, "error while receiving datagram: %m");
		continue;
	    }
	    
	    if(cs == s4) {
		if(ret < sizeof(iphdr) + sizeof(req))
		    continue;
		hdrlen = sizeof(iphdr);
		memcpy(&iphdr, buf, sizeof(iphdr));
		if(iphdr.protocol != IPPROTO_ICMP)
		    continue;
		mhdr.msg_control = NULL;
		mhdr.msg_controllen = 0;
	    } else if(cs == s6) {
		if(ret < sizeof(req))
		    continue;
		((struct sockaddr_in6 *)&name)->sin6_port = 0;
		hdrlen = 0;
		/* Just keep mhdr.msg_control. */
	    } else {
		syslog(LOG_CRIT, "strangeness!");
		abort();
	    }
	    memcpy(&req, buf + hdrlen, sizeof(req));
	    if(req.type != ICMP_NAMEREQ)
		continue;
	    rep.type = ICMP_NAMEREP;
	    rep.code = 0;
	    rep.id = req.id;
	    rep.seq = req.seq;
	    rep.ttl = htonl(ttl);
	    memcpy(buf, &rep, sizeof(rep));
	    datalen = filldn(buf + sizeof(rep));
	
	    cksum(buf, datalen + sizeof(rep));
	    
	    iov.iov_len = sizeof(rep) + datalen;
	    iov.iov_base = buf;
	    mhdr.msg_iov = &iov;
	    mhdr.msg_iovlen = 1;
	    ret = sendmsg(cs, &mhdr, 0);
	    if(ret < 0)
		syslog(LOG_WARNING, "error in sending reply: %m");
	}
    }
    
    close(s4);
    close(s6);
    return(0);
}

/*
 * Local Variables:
 * compile-command: "make CFLAGS='-Wall -g'"
 * End:
 */


src/idnlookup.c
/*
 *  idnlookup - ICMP Domain Name lookup utility for Linux
 *  Should be installed SUID root, even though I don't know if it's secure yet. :-)
 *  Copyright (C) 2005 Fredrik Tolf <fredrik@dolda2000.com>
 *  
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <netdb.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/poll.h>
#include <sys/time.h>
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

struct icmphdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
};

struct reqhdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
    u_int16_t id;
    u_int16_t seq;
};

struct rephdr {
    u_int8_t type;
    u_int8_t code;
    u_int16_t checksum;
    u_int16_t id;
    u_int16_t seq;
    int32_t ttl;
    
};

#define ICMP_NAMEREQ 37
#define ICMP_NAMEREP 38

unsigned char buf[65536];

/* DN decompression not yet implemented, since I don't know where to
 * begin counting the offset from -- the beginning of the ICMP
 * payload, or from the beginning of the DN data buffer? */
void printdn(FILE *f, unsigned char *dnbuf, size_t size, int onlyfirst)
{
    unsigned char *p;
    
    p = dnbuf;
    while(p - dnbuf < size) {
	while(*p != 0) {
	    if(*p & 0xc0) {
		fprintf(stderr, "domain name decompression not implemented, aborting\n");
		exit(1);
	    }
	    if((int)*p + (p - dnbuf) >= size) {
		fprintf(stderr, "malformed domain name\n");
		return;
	    }
	    fprintf(f, "%.*s", (int)*p, p + 1);
	    p += 1 + (int)*p;
	    if(*p != 0)
		fprintf(f, ".");
	}
	p++;
	fprintf(f, "\n");
	if(onlyfirst)
	    break;
    }
}

void cksum(void *hdr, size_t len)
{
    struct icmphdr *ih;
    u_int8_t *cb;
    int i;
    int b1, b2;
    
    ih = (struct icmphdr *)hdr;
    cb = (u_int8_t *)hdr;
    ih->checksum = 0;
    b1 = b2 = 0;
    for(i = 0; i < (len & ~1); i += 2) {
	b1 += cb[i];
	b2 += cb[i + 1];
    }
    if(i & 1)
	b1 += cb[len - 1];
    while(1) {
	if(b1 >= 256) {
	    b2 += b1 >> 8;
	    b1 &= 0xff;
	    continue;
	}
	if(b2 >= 256) {
	    b1 += b2 >> 8;
	    b2 &= 0xff;
	    continue;
	}
	break;
    }
    cb = (u_int8_t *)&ih->checksum;
    cb[0] = ~(u_int8_t)b1;
    cb[1] = ~(u_int8_t)b2;
}

void usage(void)
{
    fprintf(stderr, "usage: idnlookup [-hTa] [-t timeout] host\n");
}

int main(int argc, char **argv)
{
    int ret;
    int s, c;
    int id;
    int namelen;
    struct reqhdr req;
    struct rephdr rep;
    struct iphdr iphdr;
    size_t hdrlen;
    struct addrinfo *ai, *cai, aihint;
    struct pollfd pfd;
    struct timeval tvb, tvc;
    struct sockaddr_storage name;
    int timeout, dispttl, onlyfirst;
    int elapsed, timedout, found;
    
    timeout = 3000;
    dispttl = 0;
    onlyfirst = 1;
    while((c = getopt(argc, argv, "haTt:")) != -1) {
	switch(c) {
	case 't':
	    timeout = atoi(optarg);
	    break;
	case 'a':
	    onlyfirst = 0;
	    break;
	case 'T':
	    dispttl = 1;
	    break;
	case 'h':
	case '?':
	case ':':
	default:
	    usage();
	    exit((c == 'h')?0:1);
	}
    }
    
    if(argc - optind < 1) {
	usage();
	exit(1);
    }
    
    memset(&aihint, 0, sizeof(aihint));
    aihint.ai_socktype = SOCK_RAW;
    aihint.ai_protocol = IPPROTO_ICMP;
    ret = getaddrinfo(argv[optind], NULL, &aihint, &ai);
    
    for(cai = ai; cai != NULL; cai = cai->ai_next) {
	if((s = socket(cai->ai_family, SOCK_RAW, IPPROTO_ICMP)) < 0) {
	    perror("could not create raw socket");
	    exit(1);
	}
	
	id = random() % 65536;
	memset(&req, 0, sizeof(req));
	req.type = ICMP_NAMEREQ;
	req.id = htons(id);
	cksum(&req, sizeof(req));
	
	ret = sendto(s, &req, sizeof(req), 0, cai->ai_addr, cai->ai_addrlen);
	if(ret < 0) {
	    perror("sendto");
	    exit(1);
	} else if(ret != sizeof(req)) {
	    fprintf(stderr, "socket would not send entire packet\n");
	    exit(1);
	}
	
	timedout = 0;
	found = 0;
	gettimeofday(&tvb, NULL);
	while(1) {
	    pfd.fd = s;
	    pfd.events = POLLIN;
	    gettimeofday(&tvc, NULL);
	    elapsed = ((tvc.tv_sec - tvb.tv_sec) * 1000) + ((tvc.tv_usec - tvb.tv_usec) / 1000);
	    if(elapsed >= timeout) {
		timedout = 1;
		break;
	    }
	    ret = poll(&pfd, 1, timeout - elapsed);
	    if(ret < 0) {
		perror("idnlookup: reading data");
		exit(1);
	    }
	    
	    if(pfd.revents & POLLIN) {
		namelen = sizeof(name);
		ret = recvfrom(s, buf, sizeof(buf), 0, (struct sockaddr *)&name, &namelen);
		if(ret < 0) {
		    perror("idnlookup: receiving data");
		    exit(1);
		}
		
		if(name.ss_family != cai->ai_addr->sa_family)
		    continue;
		if(name.ss_family == AF_INET) {
		    if(memcmp(&(((struct sockaddr_in *)&name)->sin_addr), &(((struct sockaddr_in *)cai->ai_addr)->sin_addr), sizeof(struct in_addr)))
			continue;
		    if(ret < sizeof(iphdr) + sizeof(rep))
			continue;
		    hdrlen = sizeof(iphdr);
		    memcpy(&iphdr, buf, sizeof(iphdr));
		    if(iphdr.protocol != IPPROTO_ICMP)
			continue;
		} else if(name.ss_family == AF_INET6) {
		    if(memcmp(&(((struct sockaddr_in6 *)&name)->sin6_addr), &(((struct sockaddr_in6 *)cai->ai_addr)->sin6_addr), sizeof(struct in6_addr)))
			continue;
		    if(ret < sizeof(rep))
			continue;
		    hdrlen = 0;
		} else {
		    continue;
		}
		
		memcpy(&rep, buf + hdrlen, sizeof(rep));
		if(rep.type != ICMP_NAMEREP)
		    continue;
		if((ntohs(rep.id) != id) || (ntohs(rep.seq != 0)))
		    continue;
		
		found = 1;
		break;
	    }
	}
	
	close(s);
	
	if(found) {
	    if(dispttl)
		printf("%i\n", ntohl(rep.ttl));
	    printdn(stdout, buf + hdrlen + sizeof(rep), ret - hdrlen - sizeof(rep), onlyfirst);
	    break;
	}
    }
    
    if(timedout) {
	fprintf(stderr, "idnlookup: timeout\n");
	exit(1);
    }
    
    return(0);
}

/*
 * Local Variables:
 * compile-command: "make CFLAGS='-Wall -g'"
 * End:
 */


src/nss-icmp.c
/*
 *  nss-icmp or libnss_icmp - GNU C Library NSS module to query host
 *  names by ICMP.
 *  Copyright (C) 2005 Fredrik Tolf <fredrik@dolda2000.com>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Library General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Library General Public License for more details.
 *
 *  You should have received a copy of the GNU Library General Public
 *  License along with this library; if not, write to the Free
 *  Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 *  MA 02111-1307, USA
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <nss.h>
#include <sys/types.h>
#include <fcntl.h>
#include <sys/wait.h>
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef	CONFIGFILE
#define CONFIGFILE "/etc/nss-icmp.conf"
#endif

#if 0
#define DEBUGP(format...) fprintf(stderr, "nss-icmp: " format);
#else
#define DEBUGP(format...)
#endif

struct cache {
    struct cache *next, *prev;
    char *addr;
    socklen_t addrlen;
    int af;
    int notfound;
    char **names;
    time_t at, ttl;
};

static int inited = 0;
static int timeout = -1;
static int usecache = 1;
static time_t nfttl = 300;
static struct cache *cache = NULL;

static void readconfig(void)
{
    FILE *f;
    char linebuf[1024];
    char *p, *p2;
    
    if((f = fopen(CONFIGFILE, "r")) == NULL)
	return;
    
    while(fgets(linebuf, sizeof(linebuf), f) != NULL) {
	if(linebuf[0] == '#')
	    continue;
	if((p = strchr(linebuf, '\n')) != NULL)
	    *p = 0;
	if((p = strchr(linebuf, ' ')) != NULL) {
	    p2 = p + 1;
	    *p = 0;
	}
	if(!strcmp(linebuf, "timeout")) {
	    if(p2 == NULL)
		continue;
	    timeout = atoi(p2);
	}
	if(!strcmp(linebuf, "ttlnotfound")) {
	    if(p2 == NULL)
		continue;
	    nfttl = atoi(p2);
	}
	if(!strcmp(linebuf, "nocache")) {
	    usecache = 0;
	}
    }
    
    fclose(f);
}

static void freecache(struct cache *cc)
{
    int i;
    
    if(cc->next != NULL)
	cc->next->prev = cc->prev;
    if(cc->prev != NULL)
	cc->prev->next = cc->next;
    if(cc == cache)
	cache = cc->next;
    if(cc->addr != NULL)
	free(cc->addr);
    if(cc->names != NULL) {
	for(i = 0; cc->names[i] != NULL; i++)
	    free(cc->names[i]);
	free(cc->names);
    }
    free(cc);
}

static void cachenotfound(const void *addr, socklen_t len, int af, time_t ttl)
{
    struct cache *cc;
    
    for(cc = cache; cc != NULL; cc = cc->next) {
	if((cc->af == af) && (cc->addrlen == len) && !memcmp(cc->addr, addr, len))
	    break;
    }
    if(cc == NULL) {
	if((cc = malloc(sizeof(*cc))) == NULL)
	    return;
	memset(cc, 0, sizeof(*cc));
	if((cc->addr = malloc(len)) == NULL) {
	    freecache(cc);
	    return;
	}
	memcpy(cc->addr, addr, len);
	cc->addrlen = len;
	cc->af = af;
	cc->at = time(NULL);
	cc->ttl = ttl;
	
	cc->notfound = 1;
	
	cc->next = cache;
	if(cache != NULL)
	    cache->prev = cc;
	cache = cc;
    }
}

static void updatecache(const void *addr, socklen_t len, int af, char **names, time_t ttl)
{
    int i;
    struct cache *cc;
    
    for(cc = cache; cc != NULL; cc = cc->next) {
	if((cc->af == af) && (cc->addrlen == len) && !memcmp(cc->addr, addr, len))
	    break;
    }
    if(cc == NULL) {
	if((cc = malloc(sizeof(*cc))) == NULL)
	    return;
	memset(cc, 0, sizeof(*cc));
	if((cc->addr = malloc(len)) == NULL) {
	    freecache(cc);
	    return;
	}
	memcpy(cc->addr, addr, len);
	cc->addrlen = len;
	cc->af = af;
	cc->at = time(NULL);
	cc->ttl = ttl;
	
	for(i = 0; names[i] != NULL; i++);
	if((cc->names = malloc(sizeof(*(cc->names)) * (i + 1))) == NULL) {
	    freecache(cc);
	    return;
	}
	memset(cc->names, 0, sizeof(*(cc->names)) * (i + 1));
	for(i = 0; names[i] != NULL; i++) {
	    if((cc->names[i] = malloc(strlen(names[i]) + 1)) == NULL) {
		freecache(cc);
		return;
	    }
	    strcpy(cc->names[i], names[i]);
	}
	
	cc->next = cache;
	if(cache != NULL)
	    cache->prev = cc;
	cache = cc;
    }
}

static void expirecache(void)
{
    struct cache *cc, *next;
    time_t now;
    
    now = time(NULL);
    for(cc = cache; cc != NULL; cc = next) {
	next = cc->next;
	if(now - cc->at > cc->ttl) {
	    freecache(cc);
	    continue;
	}
    }
}

enum nss_status _nss_icmp_gethostbyaddr_r(const void *addr, socklen_t len, int af, struct hostent *result, char *buffer, size_t buflen, int *errnop, int *h_errnop)
{
    int i, ret;
    struct retstruct {
	char *aliaslist[16];
	char *addrlist[2];
	char retaddr[16];
    } *retbuf;
    char addrbuf[1024];
    int an, thislen, ttl;
    char *p, *p2, *p3;
    u_int8_t *ap;
    pid_t child;
    int pfd[2];
    int rl;
    int status;
    struct cache *cc;
    
    if(!inited) {
	readconfig();
	inited = 1;
    }
    
    retbuf = (struct retstruct *)buffer;
    if((buflen < sizeof(*retbuf)) || (len > sizeof(retbuf->retaddr))) {
	*errnop = ENOMEM;
	*h_errnop = NETDB_INTERNAL;
	return(NSS_STATUS_UNAVAIL);
    }
    
    DEBUGP("starting lookup\n");
    
    if(usecache) {
	expirecache();
	for(cc = cache; cc != NULL; cc = cc->next) {
	    if((cc->af == af) && (cc->addrlen == len) && !memcmp(cc->addr, addr, len))
		break;
	}
    } else {
	cc = NULL;
    }
    
    if(cc == NULL) {
	DEBUGP("address not in cache, looking up for real\n");
	ap = (u_int8_t *)addr;
	if(inet_ntop(af, addr, addrbuf, sizeof(addrbuf)) == NULL) {
	    *errnop = errno;
	    *h_errnop = NETDB_INTERNAL;
	    return(NSS_STATUS_UNAVAIL);
	}
	DEBUGP("address is %s\n", addrbuf);
    
	if(pipe(pfd)) {
	    *errnop = errno;
	    *h_errnop = NETDB_INTERNAL;
	    return(NSS_STATUS_UNAVAIL);
	}
	/* I honestly don't know if it is considered OK to fork in other
	 * people's programs. We need a SUID worker, though, so there's
	 * little choice that I can see. */
	if((child = fork()) < 0) {
	    *errnop = errno;
	    *h_errnop = NETDB_INTERNAL;
	    return(NSS_STATUS_UNAVAIL);
	}
    
	if(child == 0) {
	    int i, fd;
	    char timeoutbuf[128];
	
	    if((fd = open("/dev/null", O_WRONLY)) < 0)
		exit(127);
	    close(pfd[0]);
	    dup2(pfd[1], 1);
	    dup2(fd, 2);
	    for(i = 3; i < FD_SETSIZE; i++)
		close(i);
	
	    if(timeout != -1) {
		snprintf(timeoutbuf, sizeof(timeoutbuf), "%i", timeout);
		execlp("idnlookup", "idnlookup", "-Tt", timeoutbuf, addrbuf, NULL);
	    } else {
		execlp("idnlookup", "idnlookup", "-T", addrbuf, NULL);
	    }
	    exit(127);
	}
    
	close(pfd[1]);
    
	rl = 0;
	do {
	    ret = read(pfd[0], addrbuf + rl, sizeof(addrbuf) - rl);
	    if(ret < 0) {
		*errnop = errno;
		*h_errnop = NETDB_INTERNAL;
		close(pfd[0]);
		return(NSS_STATUS_UNAVAIL);
	    }
	    rl += ret;
	    if(rl >= sizeof(addrbuf) - 1) {
		*errnop = ENOMEM;
		*h_errnop = NETDB_INTERNAL;
		close(pfd[0]);
		return(NSS_STATUS_UNAVAIL);
	    }
	} while(ret != 0);
	addrbuf[rl] = 0;
	close(pfd[0]);
	
	waitpid(child, &status, 0);
	
	if((p = strchr(addrbuf, '\n')) == NULL) {
	    if(usecache)
		cachenotfound(addr, len, af, nfttl);
	    *h_errnop = TRY_AGAIN; /* XXX: Is this correct? */
	    return(NSS_STATUS_NOTFOUND);
	}
	*(p++) = 0;
	ttl = atoi(addrbuf);
	
	an = 0;
	p3 = buffer + sizeof(*retbuf);
	while((p2 = strchr(p, '\n')) != NULL) {
	    *p2 = 0;
	    thislen = p2 - p;
	    if(thislen == 0)
		continue;
	    if((p3 - buffer) + thislen + 1 > buflen) {
		*errnop = ENOMEM;
		*h_errnop = NETDB_INTERNAL;
		return(NSS_STATUS_UNAVAIL);
	    }
	    memcpy(p3, p, thislen + 1);
	    retbuf->aliaslist[an] = p3;
	    p3 += thislen + 1;
	    p = p2 + 1;
	    if(++an == 16) {
		*errnop = ENOMEM;
		*h_errnop = NETDB_INTERNAL;
		return(NSS_STATUS_UNAVAIL);
	    }
	}
	if(an == 0) {
	    if(usecache)
		cachenotfound(addr, len, af, nfttl);
	    *h_errnop = TRY_AGAIN; /* XXX: Is this correct? */
	    return(NSS_STATUS_NOTFOUND);
	}
	retbuf->aliaslist[an] = NULL;
	
	if(usecache)
	    updatecache(addr, len, af, retbuf->aliaslist, ttl);
    } else {
	DEBUGP("address found in cache\n");
	if(cc->notfound) {
	    *h_errnop = TRY_AGAIN; /* XXX: Is this correct? */
	    return(NSS_STATUS_NOTFOUND);
	}
	
	p3 = buffer + sizeof(*retbuf);
	for(i = 0; cc->names[i] != NULL; i++) {
	    thislen = strlen(cc->names[i]);
	    DEBUGP("filling in address %s, length %i\n", cc->names[i], thislen);
	    if((p3 - buffer) + thislen + 1 > buflen) {
		*errnop = ENOMEM;
		*h_errnop = NETDB_INTERNAL;
		return(NSS_STATUS_UNAVAIL);
	    }
	    memcpy(p3, cc->names[i], thislen + 1);
	    retbuf->aliaslist[i] = p3;
	    p3 += thislen + 1;
	}
	retbuf->aliaslist[i] = NULL;
    }
    
    DEBUGP("returning hostent\n");
    memcpy(retbuf->retaddr, addr, len);
    retbuf->addrlist[0] = retbuf->retaddr;
    retbuf->addrlist[1] = NULL;
    result->h_name = retbuf->aliaslist[0];
    result->h_aliases = retbuf->aliaslist;
    result->h_addr_list = retbuf->addrlist;
    result->h_addrtype = af;
    result->h_length = len;
    
    *h_errnop = NETDB_SUCCESS;
    DEBUGP("returning\n");
    return(NSS_STATUS_SUCCESS);
}

/*
 * Local Variables:
 * compile-command: "gcc -shared -Wall -g -o libnss_icmp.so.2 nss-icmp.c"
 * End:
 */
